aardvark/persistence/sqlalchemy/__init__.py

```diff
@@ -0,0 +1,304 @@
+                item_id=item_id,
+                lastAuthenticated=lastAuthenticated,
+                serviceName=serviceName,
+                serviceNamespace=serviceNamespace,
+                lastAuthenticatedEntity=lastAuthenticatedEntity,
+                totalAuthenticatedEntities=totalAuthenticatedEntities,
+            )
+            session.add(item)
+            return
+
+        # sqlite will return a string for item.lastAuthenticated, so we parse that into a datetime
+        if isinstance(item.lastAuthenticated, str):
+            ts = datetime.datetime.strptime(
+                item.lastAuthenticated, "%Y-%m-%d %H:%M:%S.%f"
+            )
+        else:
+            ts = item.lastAuthenticated
+
+        if lastAuthenticated > ts:
+            item.lastAuthenticated = lastAuthenticated
+            session.add(item)
+
+        elif lastAuthenticated < ts:
+            """
+            lastAuthenticated is obtained by calling get_service_last_accessed_details() method of the boto3 iam client.
+            When there is no AA data about a service, the lastAuthenticated key is missing from the returned dictionary.
+            This is perfectly valid, either because the service in question was not accessed in the past 365 days or
+            the entity granting  access to it was created recently enough that no AA data is available yet (it can take up to
+            4 hours for this to happen).
+            When this happens, the AccountToUpdate._get_job_results() method will set lastAuthenticated to 0.
+            Usually we don't want to persist such an entity, with one exception: there's already a recorded, non-zero lastAuthenticated
+            timestamp persisted for this item. That means the service was accessed at some point in time, but now more than 365 passed since
+            the last access, so AA no longer returns a timestamp for it.
+            """
+            if lastAuthenticated == 0:
+                log.warning(
+                    "Previously seen object not accessed in the past 365 days "
+                    "(got null lastAuthenticated from AA). Setting to 0. "
+                    "Object {} service {} previous timestamp {}".format(
+                        item.item_id, item.serviceName, item.lastAuthenticated
+                    )
+                )
+                item.lastAuthenticated = 0
+                session.add(item)
+            else:
+                log.error(
+                    "Received an older time than previously seen for object {} service {} ({la} < {ila})!".format(
+                        item.item_id,
+                        item.serviceName,
+                        la=lastAuthenticated,
+                        ila=item.lastAuthenticated,
+                    )
+                )
+
+    def get_or_create_iam_object(self, arn):
+        with self.session_scope() as session:
+            item = session.query(AWSIAMObject).filter(AWSIAMObject.arn == arn).scalar()
```

Comment:
Any need to catch DB errors?

Reply:
Errors here will be caught in the `session_scope()` context manager, but I'll add more granular handling and a re-raise to provide more context.